python/ray/tune/ray_trial_executor.py

```diff
@@ -252,27 +280,69 @@ def _setup_remote_runner(self, trial):
                     "implemented and return True.")
             return existing_runner
 
-        if self._cached_actor:
+        if self._cached_actor_pg[0]:
             logger.debug("Cannot reuse cached runner {} for new trial".format(
-                self._cached_actor))
-            with self._change_working_directory(trial):
-                pg = self._pg_manager.clean_trial_placement_group(trial)
+                self._cached_actor_pg[0]))
+            existing_runner, pg = self._cached_actor_pg
+
+            if not self._pg_manager.return_cached_pg(pg, replace_pending=True):
+                # If we could not replace a pending pg, it might not be needed
+                # anymore. Remove instead.
+                self._pg_manager.clean_cached_pg(pg)
+            else:
+                pg = None
```

Comment:
so ideally, we can do something like:

```python
self._pg_manager.return_cached_pg(pg)
with self._change_working_directory(trial):
    self._trial_cleanup.add(trial, actor=existing_runner)
```

a couple notes here:

1. I'm making the assumption that pg's are cleaned up at some other point in the code (see other comment in pbt.py): https://github.com/ray-project/ray/pull/13906/files#r574932606
2. return_cached_pg seems to always require replacement_pending=True anyways, and is called nowhere else in the code, so you can probably fuse the `clean_cached_pg` in.
3. Since the pg is cleaned up somewhere else in the code (maybe somewhere where you check for pgs that are 1. empty and 2. will not be used), you can then disassociate the trial cleanup from PG cleanup.

Reply:
Right, that's a great suggestion. I moved the placement group lifecycle management into the loop, and we're now reconciling the expected number of placement groups with the actual number at the end of each `step()`. This means that we usually don't remove any resources until the end of a training step and thus that external components need not to worry about this. That's much better, thanks for your help on this!